vj4/model/adaptor/contest.py

```diff
@@ -160,15 +168,27 @@ def get_multi_status(*, fields=None, **kwargs):
   if 'attend' not in tsdoc or not tsdoc['attend']:
     raise error.ContestNotAttendedError(domain_id, tid, uid)
 
-  # Sort and uniquify journal of the contest status document, by rid.
-  key_func = lambda j: j['rid']
-  journal = [list(g)[-1]
-             for _, g in itertools.groupby(sorted(tsdoc['journal'], key=key_func), key=key_func)]
+  journal = _get_status_journal(tsdoc)
   stats = RULES[tdoc['rule']].stat_func(tdoc, journal)
   tsdoc = await document.rev_set_status(domain_id, document.TYPE_CONTEST, tid, uid, tsdoc['rev'],
                                         journal=journal, **stats)
   return tsdoc
 
 
+@argmethod.wrap
+async def recalc_contest_status(domain_id: str, tid: objectid.ObjectId):
+  tdoc = await document.get(domain_id, document.TYPE_CONTEST, tid)
+  async with document.get_multi_status(domain_id=domain_id,
+                                       doc_type=document.TYPE_CONTEST,
+                                       doc_id=tdoc['doc_id']) as cursor:
+    async for tsdoc in cursor:
+      if 'attend' not in tsdoc or not tsdoc['attend']:
+        continue
+      journal = _get_status_journal(tsdoc)
+      stats = RULES[tdoc['rule']].stat_func(tdoc, journal)
+      await document.rev_set_status(domain_id, document.TYPE_CONTEST, tid, tsdoc['uid'], tsdoc['rev'],
```

Comment:
Try to use bulk: https://github.com/vijos/vj4/blob/master/vj4/job/record.py#L107

Reply:
bulk needs more memory than cursor